asdf/resource.py

```diff
@@ -0,0 +1,180 @@
+        Root directory of the resource files.  `str` will
+        be interpreted as a filesystem path.
+    uri_prefix : str
+        Prefix used to construct URIs from file paths.  The
+        prefix will be prepended to paths relative to the root
+        directory.
+    recursive : bool, optional
+        If `True`, recurse into subdirectories.  Defaults to `False`.
+    filename_pattern : str, optional
+        Glob pattern that identifies relevant filenames.
+        Defaults to "*.yaml".
+    stem_filename : bool, optional
+        If `True`, remove the filename's extension when
+        constructing its URI.
+    """
+    def __init__(self, root, uri_prefix, recursive=False, filename_pattern="*.yaml", stem_filename=True):
+        self._uri_to_traversable = {}
+        self._recursive = recursive
+        self._filename_pattern = filename_pattern
+        self._stem_filename = stem_filename
+
+        if isinstance(root, str):
+            self._root = Path(root)
+        else:
+            self._root = root
+
+        if uri_prefix.endswith("/"):
+            self._uri_prefix = uri_prefix[:-1]
+        else:
+            self._uri_prefix = uri_prefix
+
+        for traversable, path in self._iterate_files(self._root, []):
+            self._uri_to_traversable[self._make_uri(traversable, path)] = traversable
+
+    def _iterate_files(self, traversable, path):
+        for obj in traversable.iterdir():
+            if obj.is_file() and fnmatch.fnmatch(obj.name, self._filename_pattern):
+                yield obj, path
+            elif obj.is_dir() and self._recursive:
+                yield from self._iterate_files(obj, path + [obj.name])
+
+    def _make_uri(self, traversable, path):
+        if self._stem_filename:
+            filename = os.path.splitext(traversable.name)[0]
+        else:
+            filename = traversable.name
+
+        return "/".join([self._uri_prefix] + path + [filename])
+
+    def __getitem__(self, uri):
+        return self._uri_to_traversable[uri].read_bytes()
+
+    def __len__(self):
+        return len(self._uri_to_traversable)
+
+    def __iter__(self):
+        yield from self._uri_to_traversable
+
+    def __repr__(self):
+        return "{}({!r}, {!r}, recursive={!r}, filename_pattern={!r}, stem_filename={!r})".format(
+            self.__class__.__name__,
+            self._root,
+            self._uri_prefix,
+            self._recursive,
+            self._filename_pattern,
+            self._stem_filename,
+        )
+
+
+class ResourceManager(Mapping):
+    """
+    Wraps multiple resource mappings into a single interface
+    with some friendlier error handling.
+
+    Parameters
+    ----------
+    resource_mappings : list of collections.abc.Mapping
+        Underlying resource mappings.  In the case of
+        a duplicate URI, the latest mapping in the list
+        will override.
+    """
+    def __init__(self, resource_mappings):
+        self._resource_mappings = resource_mappings
+
```

Comment:
It might be good to explain why the two level mechanism has to be retained. In other words, why not create a dictionary keyed by URI to the end object immediately? Is the idea to allow changing the individual mappings in some way after the fact?

Reply:
My thought was to avoid loading the bytes until needed, in case the resources are being requested over HTTP or are otherwise slow to access.  Most resources won't be needed in a given session so loading them all up-front is unnecessary and possibly time consuming.  Implemented this way, only the URIs are read on load.